src/plugin/file/shp/shp.js

```diff
@@ -67,3 +73,123 @@ export const isSHPFileType = function(content) {
     return false;
   }
 };
+
+/**
+ * Flattens geometry coordinates down to an array of coordinate groups.
+ * @param {!Array} coordinates The current coordinates.
+ * @return {!Array<!Array<number>>} The coordinate groups.
+ */
+export const getFlatGroupCoordinates = (coordinates) => {
+  // We're trying to collapse down to an array of arrays of numbers (ie, array of coordinates).
+  while (Array.isArray(coordinates) && Array.isArray(coordinates[0]) && Array.isArray(coordinates[0][0])) {
+    coordinates = coordinates.flat(1);
+  }
+
+  return coordinates;
+};
+
+/**
+ * From an OpenLayers geometry, get the array of SHP parts for that geometry type.
+ * @param {SimpleGeometry} geometry The geometry.
+ * @return {!Array} The SHP parts for the geometry.
+ */
+export const getPartCoordinatesFromGeometry = (geometry) => {
+  let partCoordinates;
+
+  if (geometry) {
+    const coordinates = geometry.getCoordinates();
+    const geomType = geometry.getType();
+
+    //
+    // OpenLayers structures geometry coordinates from getCoordinates as:
+    //  - Point: The coordinate
+    //  - MultiPoint: Array of coordinates
+    //  - LineString: Array of coordinates
+    //  - MultiLineString: Array of lines -> coordinates
+    //  - Polygon: Array of rings -> coordinates
+    //  - MultiPolygon: Array of polygons -> rings -> coordinates
+    //
+    // SHP handles multi geometries with a "parts" value that specifies the number of points/lines/rings. It also only
+    // has a distinct shape type for MultiPoint, both line and polygon use the same shape whether there is a single geom
+    // or multiple. The following converts OpenLayers coordinates to SHP parts for all supported geometry types.
+    //
+    switch (geomType) {
+      case GeometryType.POINT:
+        // Each part is a point, add a level.
+        partCoordinates = [coordinates];
+        break;
+      case GeometryType.MULTI_POINT:
+        // Each part is a point, all set.
+        partCoordinates = coordinates;
+        break;
+      case GeometryType.LINE_STRING:
+        // Each part is a line, add a level.
+        partCoordinates = [coordinates];
+        break;
+      case GeometryType.MULTI_LINE_STRING:
+        // Each part is a line, all set.
+        partCoordinates = coordinates;
+        break;
+      case GeometryType.POLYGON:
+        // Each part is a ring, all set.
+        partCoordinates = coordinates;
+        break;
+      case GeometryType.MULTI_POLYGON:
+        // Each part is a ring, flatten by one level.
+        partCoordinates = coordinates.flat(1);
+        break;
+      default:
+        // Unsupported geometry type.
```

Comment:
I know we don't expect this to be hit, but it would probably be good to log a warning or an error just in case something weird happens.

Reply:
For the purpose of this function I think it's fine not to log anything, but the caller may want to log if an empty array is returned.